### ISSUE TITLE
Consider locating .pylintrc

### ISSUE THREAD

**Issue by silverwind** [SublimeLinter/SublimeLinter-pylint#57]
`pylint` is not very smart about finding a `.pylintrc` file. Given below directory structure and `projects` being the folder loading into ST, `file.py` will not lint using `.pylintrc`, but instead fallback to the global config:

````
projects
└─ project
   ├── .pylintrc
   └── file.py
````

You can also reproduce the issue via `pylint projects/file.py` while in the `projects` directory. It looks like  pylint only takes `.pylintrc` into account when it is in `$PWD` which is not the case in this example.

I suggest to add a small mechanism to this plugin that walks up the directory tree from the linted file to locate a rcfile and if found supply it via `--rcfile` argument, overwritable via user config.

I did raise this issue in https://github.com/PyCQA/pylint/issues/3279, but I don't think they will implement a smarter rcfile resolution anytime soon.

**Comment by kaste** [comment]
Hi 👋! I think the better approach would be to find the project root, and then just set the working dir to that. For your example, the working dir would be `projects/project` when linting `file.py`.  Since we already have a setting for `working_dir`, we would just provide a better default for it.  Any user can set this setting, effectively disabling our heuristic.

**Comment by silverwind** [comment]
> Since we already have a setting for working_dir, we would just provide a better default for it

I could see some heuristics to determine the working directory of a Python project. There's probably no surefire way, but the presence of any of these files would be strong hint:

- pyproject.toml
- requirements.txt
- setup.py

Can `working_dir` be set per-project?

**Comment by silverwind** [comment]
Another idea. Find the closest rcfile and `os.chdir` into it before linting. This will help this code find the right file:

https://github.com/PyCQA/pylint/blob/6d818e3b84b853fe06f4199eb8408fe6dfe033be/pylint/config.py#L112-L121

**Comment by kaste** [comment]
Yes, `working_dir` can be set per project.  `working_dir` is basically the idea of `chdir` for parallel, multi-threaded running of linter processes.

**Comment by silverwind** [comment]
Mind showing how one can set it? Documentation seems a bit sparse. I guess this would be a good enough solution for me now, but I still keep my hopes up that `pylint` will eventually solve it. I totally understand if you don't want to add such hacks as above to this project.

**Comment by silverwind** [comment]
Here's what I've tried so far, both don't seem to convince the plugin to use a different `working_dir`:

1. Add `.sublime_project` with 

````json
{
  "settings": {
    "SublimeLinter.linters.pylint.working_dir": "${project_path}"
  }
}
````

2. Edit global `linters`:

````json
"linters": {
  "pylint": {
    "working_dir": "${project_path}"
  }
}
````

Any pointers?

**Comment by kaste** [comment]
That's generally how you do this. Maybe not `${project_path}` but `${folder}` is what you want?  Probably turn on debug mode, then it will print the cwd (among all the other stuff), just to be sure you edit the correct file.  (It really depends on which folder(s) are the top folders you see in the sidebar.

Generally, the node based linters (typically eslint :grin:) resolve or provide a context variable `project_root`, usually the path with the nearest packages.json in it, and then the working dir defaults to this `project_root`.  I think we should/could provide this for python linters as well.

**Comment by kaste** [comment]
See https://github.com/dense-analysis/ale/blob/89db85121c001fc60787647f012978a2328816a5/autoload/ale/python.vim#L19-L64 for how this is done for vim.

**Comment by silverwind** [comment]
Nope, neither `${folder}` nor `${project_path}` make any difference, it always uses my top-level folder while the actual project root is 2 levels deeper:

```
SublimeLinter: linter.py:1103         pylint: Checking lint mode 'load_save' vs lint reason 'on_modified'.  Skip.
SublimeLinter: sublime_linter.py:324  Delay linting 'test.py' for 0.0s
SublimeLinter: linter.py:1103         pylint: Checking lint mode 'load_save' vs lint reason 'on_save'.  Ok.
SublimeLinter: sublime_linter.py:379  Friendly terminate: <pid 23281>
SublimeLinter: sublime_linter.py:575  Linting 'test.py' took 0.49s
SublimeLinter: #2 linter.py:1120      pylint: linting 'test.py'
SublimeLinter: #2 python_linter.py:44 pylint: wanted python is 'None'
SublimeLinter: #2 python_linter.py:95 pylint: trying to use globally installed pylint
SublimeLinter: #2 linter.py:1636      Running ...

  /Users/silverwind/git  (working dir)
```

**Comment by silverwind** [comment]
> how this is done for vim

That seems like a start. I'd certainly add `pyproject.toml` to the list, it's the equivalent of `package.json` when using the `poetry` package manager.

**Comment by kaste** [comment]
Working with a multi-mono-repo is always a bit tricky. You basically have to add each sub-repo as a folder to the sidebar.  Only then you can use `$folder` which will select the correct folder the file is in (searching top to bottom, so sorting the folders in the project comes to play as well).  

For an automatic approach, implementing this in SublimeLinter core is probably the right place. You can look at the `node_linter.py` over there.

**Comment by silverwind** [comment]
@kaste would you accept a PR that searches for a project root indicated by any of the mentioned files and `cd` into that directory before running `pylint`?

**Comment by kaste** [comment]
I'm positive and don't think this will go crazy code-wise.  But remember we don't `cd` here, you just search for a `path` and set `self.context['project_root'] = path`, and everything will follow magically 🤣

**Comment by kaste** [comment]
We also have `path_upwards` helper functions https://github.com/SublimeLinter/SublimeLinter/blob/master/lint/base_linter/node_linter.py#L21-L39 and their usage https://github.com/SublimeLinter/SublimeLinter/blob/master/lint/base_linter/node_linter.py#L109-L114  It probably isn't directly rocket science.

**Comment by mortenfyhn** [comment]
Just stopping by to say that something like traversing up dirs and picking the nearest .pylintrc would be great. I guess that's similar to how e.g. clang-format looks for a .clang-format file.
